### PR TITLE
dask: Deprecate `close`

### PR DESCRIPTION
**Pull request by davidhassell**
Deprecates `close` methods, and removes the corresponding tests.

### CODE REVIEW

File: cf/constructs.py [hunk 1]
```diff
@@ -131,10 +135,13 @@ def close(self):
...
-        # TODODASKAPI - is this method still needed?
-
-        for construct in self.filter_by_data().values():
-            construct.close()
+        _DEPRECATION_ERROR_METHOD(
+            self,
+            "close",
+            "All files are now automatically closed when not being accessed.",
+            version="TODODASKVER",
+            remove_at="5.0.0",
```
Comment (sadielbartholomew) [inline cf/constructs.py:143]:
```suggestion
            removed_at="5.0.0",
```

File: cf/coordinatereference.py [hunk 2]
```diff
@@ -287,7 +290,13 @@ def close(self):
         >>> c.close()
 
         """
-        pass
+        _DEPRECATION_ERROR_METHOD(
+            self,
+            "close",
+            "All files are now automatically closed when not being accessed.",
+            version="TODODASKVER",
+            remove_at="5.0.0",
```
Comment (sadielbartholomew) [inline cf/coordinatereference.py:298]:
```suggestion
            removed_at="5.0.0",
```

File: cf/domain.py [hunk 3]
```diff
@@ -139,9 +142,13 @@ def close(self):
...
         """
-        # TODODASK - is this still needed?
-
-        self.constructs.close()
+        _DEPRECATION_ERROR_METHOD(
+            self,
+            "close",
+            "All files are now automatically closed when not being accessed.",
+            version="TODODASKVER",
+            remove_at="5.0.0",
```
Comment (sadielbartholomew) [inline cf/domain.py:150]:
```suggestion
            removed_at="5.0.0",
```

File: cf/field.py [hunk 4]
```diff
@@ -4983,10 +4986,13 @@ def close(self):
...
-        super().close()
-
-        for construct in self.constructs.filter_by_data(todict=True).values():
-            construct.close()
+        _DEPRECATION_ERROR_METHOD(
+            self,
+            "close",
+            "All files are now automatically closed when not being accessed.",
+            version="TODODASKVER",
+            remove_at="5.0.0",
```
Comment (sadielbartholomew) [inline cf/field.py:4994]:
```suggestion
            removed_at="5.0.0",
```

File: cf/mixin/propertiesdata.py [hunk 5]
```diff
@@ -2515,9 +2518,13 @@ def close(self):
...
         """
-        data = self.get_data(None, _fill_value=False)
-        if data is not None:
-            data.close()
+        _DEPRECATION_ERROR_METHOD(
+            self,
+            "close",
+            "All files are now automatically closed when not being accessed.",
+            version="TODODASKVER",
+            remove_at="5.0.0",
```
Comment (sadielbartholomew) [inline cf/mixin/propertiesdata.py:2526]:
```suggestion
            removed_at="5.0.0",
```

File: cf/mixin/propertiesdatabounds.py [hunk 6]
```diff
@@ -1286,15 +1289,13 @@ def close(self):
...
-
-        interior_ring = self.get_interior_ring(None)
-        if interior_ring is not None:
-            interior_ring.close()
+        _DEPRECATION_ERROR_METHOD(
+            self,
+            "close",
+            "All files are now automatically closed when not being accessed.",
+            version="TODODASKVER",
+            remove_at="5.0.0",
```
Comment (sadielbartholomew) [inline cf/mixin/propertiesdatabounds.py:1297]:
```suggestion
            removed_at="5.0.0",
```

**Review by sadielbartholomew — COMMENTED** [review]
Close, but no cigar! :laughing: Feel free to apply the recurrent change suggested yourself rather than via my 'suggestion' set if that's more convenient, but I've made them as suggestions in case that is simplest.

Is there a location we can put some brief testing on deprecated keywords such as this, since nothing failed in the test suite despite the consistent invalid keyword argument here, and it would be good to catch basic issues like that?

**Review by sadielbartholomew — CHANGES_REQUESTED** [review]
[Intended to set this as my judgement in the review I just submitted...]

**Comment by davidhassell** [comment]
> Is there a location we can put some brief testing on deprecated keywords such as this, since nothing failed in the test suite despite the consistent invalid keyword argument here, and it would be good to catch basic issues like that?

Hi Sadie, good idea. Not yet - I suppose a new file `test_deprecated.py` might be a good place to put them, from where they can easily be deleted once the `removed_at` version is reached.

**Comment by sadielbartholomew** [comment]
> I suppose a new file test_deprecated.py might be a good place to put them, from where they can easily be deleted once the removed_at version is reached.

Sounds like a good plan to me. Shall we bump it to a new issue, since the Dask migration takes priority I am sure?

**Review by sadielbartholomew — APPROVED** [review]
Feedback all addressed. All good now, please merge.

**Pull request merged**